src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyName.cs

```diff
@@ -303,175 +305,92 @@ public static bool ReferenceMatchesDefinition(AssemblyName? reference, AssemblyN
-                            throw new FormatException(SR.Arg_FormatException);
+                        r = Rune.ReplacementChar;
+                    }
 
-                        i += (count - 1);
+                    Debug.Assert(stringToEscape.EnumerateRunes() is { } e && e.MoveNext() && e.Current == r);
+                    Debug.Assert(charsConsumed is 1 or 2);
 
-                        for (count = 0; count < numberOfBytes; ++count)
-                            EscapeAsciiChar((char)bytes[count], dest, ref destPos);
+                    stringToEscape = stringToEscape.Slice(charsConsumed);
 
-                        prevInputPos = i + 1;
-                    }
-                    else if (ch == '%' && rsvd == '%')
-                    {
-                        // Means we don't reEncode '%' but check for the possible escaped sequence
-                        dest = EnsureDestinationSize(pStr, dest, i, c_EncodedCharsPerByte,
-                            c_MaxAsciiCharsReallocate * c_EncodedCharsPerByte, ref destPos, prevInputPos);
-                        if (i + 2 < end && char.IsAsciiHexDigit(pStr[i + 1]) && char.IsAsciiHexDigit(pStr[i + 2]))
-                        {
-                            // leave it escaped
-                            dest[destPos++] = '%';
-                            dest[destPos++] = pStr[i + 1];
-                            dest[destPos++] = pStr[i + 2];
-                            i += 2;
-                        }
-                        else
-                        {
-                            EscapeAsciiChar('%', dest, ref destPos);
-                        }
-                        prevInputPos = i + 1;
-                    }
-                    else if (ch == force1 || ch == force2 || (ch != rsvd && (isUriString ? !IsReservedUnreservedOrHash(ch) : !IsUnreserved(ch))))
+                    // The rune is non-ASCII, so encode it as UTF8, and escape each UTF8 byte.
+                    r.TryEncodeToUtf8(utf8Bytes, out int bytesWritten);
+                    foreach (byte b in utf8Bytes.Slice(0, bytesWritten))
                     {
-                        dest = EnsureDestinationSize(pStr, dest, i, c_EncodedCharsPerByte,
-                            c_MaxAsciiCharsReallocate * c_EncodedCharsPerByte, ref destPos, prevInputPos);
-                        EscapeAsciiChar(ch, dest, ref destPos);
-                        prevInputPos = i + 1;
+                        PercentEncodeByte(b, ref vsb);
                     }
                 }
-
-                if (prevInputPos != i)
+                else if (!UnreservedReserved.Contains(ch))
                 {
-                    // need to fill up the dest array ?
-                    if (prevInputPos != start || dest != null)
-                        dest = EnsureDestinationSize(pStr, dest, i, 0, 0, ref destPos, prevInputPos);
+                    PercentEncodeByte((byte)ch, ref vsb);
+                    stringToEscape = stringToEscape[1..];
```

Comment:
It's possible it's improved since, but last I checked it was still slightly better to just do `stringToEscape.Slice(1)`

Reply:
I thought I had undone all of Roslyn's suggestions. Fixed, thanks!